src/Css/Style.php

```diff
@@ -1220,6 +1220,39 @@ private function evaluate_func_calc(array $rpn, float $ref_size = 0, ?float $fon
         return floatval(end($stack));
     }
 
+    /**
+     * Resolves the actual values for used CSS custom properties.
+     *
+     * This function receives the whole content of the var() function, which
+     * can also include a fallback value.
+     */
+    private function parse_var(string $variable) {
+        // Split property name and an optional fallback value.
+        [$custom_prop, $fallback] = explode(',', $variable, 2) + [null, null];
+        $fallback = trim($fallback);
+
+        // Recursively try to resolve the property value through all parents.
+        $value = $this->_props[$custom_prop] ?? (
+            isset($this->parent_style)
+                ? $this->parent_style->parse_var($variable)
+                : null
+        );
+
+        // If the resolved value is a variable, resolve that too.
+        if (is_string($value) && substr($value, 0, 4) === 'var(') {
+            $var = substr(trim($value), 4, -1);
+            $value = $this->parse_var($var);
+        }
+
+        // If the resolved fallback is a variable, resolve that too.
+        if (is_string($fallback) && substr($fallback, 0, 4) === 'var(') {
+            $var = substr(trim($fallback), 4, -1);
+            $fallback = $this->parse_var($var);
+        }
```

Comment:
Relevant section from the spec: https://www.w3.org/TR/css-variables-1/#cycles

Reply:
Depending on how the dependency issue will be solved, we might be able to use the dependency map to detect circulair reference?

Reply:
The more I think of it, the less it makes sense to work with dependencies. The whole fallback thing is there because a custom property might not have been defined at all. Using with dependencies will not work and properties that have a falid fallback might get dropped as invalid/incomplete. 

I suppose the simplest way to keep the recursion under control is adding something like this?

    if (count(debug_backtrace()) > 99) {
       // Add error message to global var.
       return;
    }

Or you want to add the max depth as an option in Dompdf/Options? PHP does not enfore a max recursion depth, though Xdebug stops at 100. Can we assume that people test their implementation? i.e. The performance impact of stopping the recursion relatively late should not impact a production system? 
If we stop at a lower version to save performance we might break a valid setup in a framework with lots of calls?

Reply:
For the recursion check I'm not using a dependency map. I added a stack to track which variables have been processed. If we encounter the same custom property reference more than once the method bails. I'm still thinking through fallback handling before I check something in for review.

Reply:
Oké, Thanks. I know we spoke about dependencies in other threads as well, but this was the most relevant open one.

I just suddenly realized how strict dependency checking could result in unexpected behavior for undefined properties, which I wanted to write down somewhere.

Reply:
The spec does dictate a "directed dependency graph" which this is most certainly not. Going that route would require a bit more effort that I'm able to dedicate at the moment. Though that level of implementation would probably provide some relief from the issues I'm trying to work around right now.